luomu-common/src/directed_addr.rs

```diff
@@ -0,0 +1,106 @@
+/// A `Source` of any kind. For example an IP address.
+#[derive(Clone, Copy, Debug, Hash, PartialEq, Eq)]
+pub struct Source<ADDR>(ADDR);
+
+impl<ADDR> Source<ADDR> {
+    /// Constructs a new `Source`.
+    pub const fn new(addr: ADDR) -> Self {
+        Self(addr)
+    }
+
+    /// Make a [Destination] out from `Source`.
+    pub fn flip(self) -> Destination<ADDR> {
+        Destination(self.unwrap())
+    }
+
+    /// Returns the underlying value inside `Source`.
+    pub fn unwrap(self) -> ADDR {
+        self.0
+    }
+}
+
+/// A `Destination` of any kind. For example an IP address.
+#[derive(Clone, Copy, Debug, Hash, PartialEq, Eq)]
+pub struct Destination<ADDR>(ADDR);
+
+impl<ADDR> Destination<ADDR> {
```

Comment:
Just thinking aloud here, would it be useful to add `Impl Deref` which would return the underlying address (and thus just call `unwrap()` ). 

Obviously same for `Source` as well

Reply:
Implemented in 40e99b38fb738887b376e18e9aef6ee2143579c5. I added `Deref`, `DerefMut`, `AsRef` and `AsMut`, and couple `From` trait implementations.